libraries/shared/storage/backfiller.go

```diff
@@ -0,0 +1,165 @@
+// VulcanizeDB
+// Copyright © 2019 Vulcanize
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package storage
+
+import (
+	"errors"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/statediff"
+	"github.com/sirupsen/logrus"
+
+	"github.com/vulcanize/vulcanizedb/libraries/shared/fetcher"
+	"github.com/vulcanize/vulcanizedb/libraries/shared/storage/utils"
+)
+
+const (
+	DefaultMaxBatchSize   uint64 = 1000
+	defaultMaxBatchNumber int64  = 10
+)
+
+// BackFiller is the backfilling interface
+type BackFiller interface {
+	BackFill(endingBlock uint64, backFill chan utils.StorageDiff, errChan chan error, done chan bool) error
+}
+
+// backFiller is the backfilling struct
+type backFiller struct {
+	fetcher       fetcher.StateDiffFetcher
+	batchSize     uint64
+	startingBlock uint64
+}
+
+// NewStorageBackFiller returns a BackFiller
+func NewStorageBackFiller(fetcher fetcher.StateDiffFetcher, startingBlock, batchSize uint64) BackFiller {
+	if batchSize == 0 {
+		batchSize = DefaultMaxBatchSize
+	}
+	return &backFiller{
+		fetcher:       fetcher,
+		batchSize:     batchSize,
+		startingBlock: startingBlock,
+	}
+}
+
+// BackFill fetches, processes, and returns utils.StorageDiffs over a range of blocks
+// It splits a large range up into smaller chunks, batch fetching and processing those chunks concurrently
+func (bf *backFiller) BackFill(endingBlock uint64, backFill chan utils.StorageDiff, errChan chan error, done chan bool) error {
+	if endingBlock < bf.startingBlock {
+		return errors.New("backfill: ending block number needs to be greater than starting block number")
+	}
+	logrus.Infof("going to fill in gap from %d to %d", bf.startingBlock, endingBlock)
+	// break the range up into bins of smaller ranges
+	length := endingBlock - bf.startingBlock + 1
+	numberOfBins := length / bf.batchSize
+	remainder := length % bf.batchSize
+	if remainder != 0 {
+		numberOfBins++
+	}
+	blockRangeBins := make([][]uint64, numberOfBins)
+	for i := range blockRangeBins {
+		nextBinStart := bf.startingBlock + uint64(bf.batchSize)
+		if nextBinStart > endingBlock {
+			nextBinStart = endingBlock + 1
```

Comment:
What's the purpose of mutating `nextBinStart` in this manner? Are we trying to avoid iterating through block numbers past the ending block? Would it be possible to achieve a similar outcome by applying that condition on line 82? e.g. `for j := bf.startingBlock; j < nextBinStart && j < endingBlock; j++`

Reply:
`j <= endingBlock` but yes!